### PR TITLE
support for KType

### PR DESCRIPTION
**Pull request by evgeny-pasynkov**
Support for building TypeNames from kotlin.reflect.KType (#325 )

### CODE REVIEW

File: src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt [hunk 1]
```diff
@@ -108,9 +113,40 @@ class ParameterizedTypeName internal constructor(
...
 
 /** Returns a parameterized type equivalent to `type`.  */
 @JvmName("get")
 fun ParameterizedType.asParameterizedTypeName() = ParameterizedTypeName.get(this, mutableMapOf())
+
+/** Returns a class name equivalent to given Kotlin KType.  */
+fun KType.asClassName(): TypeName {
+  val classifier = this.classifier
+  if (classifier is KTypeParameter)
+    return ClassName("", classifier.name)
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt]: This doesn't look correct. Type parameters are not classes.

File: src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt [hunk 2]
```diff
@@ -108,9 +113,40 @@ class ParameterizedTypeName internal constructor(
...
 }
 
 /** Returns a parameterized type equivalent to `type`.  */
 @JvmName("get")
 fun ParameterizedType.asParameterizedTypeName() = ParameterizedTypeName.get(this, mutableMapOf())
+
+/** Returns a class name equivalent to given Kotlin KType.  */
+fun KType.asClassName(): TypeName {
+  val classifier = this.classifier
+  if (classifier is KTypeParameter)
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt]: nit: missing `{` and `}`

File: src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt [hunk 3]
```diff
@@ -108,9 +113,40 @@ class ParameterizedTypeName internal constructor(
...
+    return ClassName("", classifier.name)
+
+  if (classifier == null || classifier !is KClass<*>)
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt]: nit: missing `{` and `}`

File: src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt [hunk 4]
```diff
@@ -108,9 +113,40 @@ class ParameterizedTypeName internal constructor(
         get(ownerType, map = map).nestedClass(rawType.simpleName(), typeArguments) else
         ParameterizedTypeName(null, rawType, typeArguments)
     }
+
+    /** Returns a type name equivalent to type with given list of type arguments.  */
+    internal fun get(type: KClass<*>, nullable: Boolean, typeArguments: List<KTypeProjection>): TypeName {
+      if (typeArguments.isEmpty())
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt]: nit: missing `{` and `}`

File: src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt [hunk 5]
```diff
@@ -108,9 +113,40 @@ class ParameterizedTypeName internal constructor(
...
+        return type.asTypeName()
+
+      val enclosingClass = type.java.enclosingClass?.kotlin
+
+      return ParameterizedTypeName(
+              enclosingClass?.let {
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt]: continuation indent is 4 spaces not 8

File: src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt [hunk 6]
```diff
@@ -80,26 +84,27 @@ class ParameterizedTypeName internal constructor(
...
     internal fun get(
-      type: ParameterizedType,
-      map: MutableMap<Type, TypeVariableName>
+            type: ParameterizedType,
+            map: MutableMap<Type, TypeVariableName>
     ): ParameterizedTypeName {
       val rawType = (type.rawType as Class<*>).asClassName()
       val ownerType = if (type.ownerType is ParameterizedType
-          && !Modifier.isStatic((type.rawType as Class<*>).modifiers))
+              && !Modifier.isStatic((type.rawType as Class<*>).modifiers))
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt]: Revert spacing changes

File: src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt [hunk 7]
```diff
@@ -80,26 +84,27 @@ class ParameterizedTypeName internal constructor(
...
+    @JvmStatic
+    fun get(rawType: KClass<*>, vararg typeArguments: KClass<*>) = ParameterizedTypeName(null, rawType.asClassName(),
+            typeArguments.map { it.asTypeName() })
 
     /** Returns a parameterized type, applying `typeArguments` to `rawType`.  */
-    @JvmStatic fun get(rawType: Class<*>, vararg typeArguments: Type) = ParameterizedTypeName(
-        null, rawType.asClassName(), typeArguments.map { it.asTypeName() })
+    @JvmStatic
+    fun get(rawType: Class<*>, vararg typeArguments: Type) = ParameterizedTypeName(
+            null, rawType.asClassName(), typeArguments.map { it.asTypeName() })
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt]: Can you revert all these reformats?

File: src/test/java/com/squareup/kotlinpoet/AbstractTypesTest.kt [hunk 8]
```diff
@@ -182,6 +186,12 @@ abstract class AbstractTypesTest {
     assertThat(type.toString()).isEqualTo("T") // (Bounds are only emitted in declaration.)
   }
 
+  @Test fun kType() {
+    assertThat(Map::class.starProjectedType.asClassName().toString()).isEqualTo("kotlin.collections.Map<*, *>")
+    assertThat(Map::class.createType(listOf(KTypeProjection(KVariance.INVARIANT, String::class.createType(emptyList())), KTypeProjection.STAR)).asClassName().toString()).isEqualTo("kotlin.collections.Map<kotlin.String, *>")
+    assertThat(Map.Entry::class.createType(listOf(KTypeProjection(KVariance.INVARIANT, String::class.createType(emptyList())), KTypeProjection.STAR)).asClassName().toString()).isEqualTo("kotlin.collections.Map.Entry<kotlin.String, *>")
```
Comment (JakeWharton) [inline src/test/java/com/squareup/kotlinpoet/AbstractTypesTest.kt]: Line wrapping?

File: src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt [hunk 9]
```diff
@@ -108,9 +112,43 @@ class ParameterizedTypeName internal constructor(
...
+
+/** Returns a class name equivalent to given Kotlin KType.  */
+fun KType.asTypeName(): TypeName {
+  val classifier = this.classifier
+  if (classifier is KTypeParameter) {
+    return classifier.asTypeVariableName()
+  }
+
+  if (classifier == null || classifier !is KClass<*>) {
+    throw IllegalArgumentException("Cannot build ClassName for $this")
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt]: s/ClassName/TypeName/

File: src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt [hunk 10]
```diff
@@ -108,9 +112,43 @@ class ParameterizedTypeName internal constructor(
...
 fun ParameterizedType.asParameterizedTypeName() = ParameterizedTypeName.get(this, mutableMapOf())
+
+/** Returns a class name equivalent to given Kotlin KType.  */
+fun KType.asTypeName(): TypeName {
+  val classifier = this.classifier
+  if (classifier is KTypeParameter) {
+    return classifier.asTypeVariableName()
+  }
+
+  if (classifier == null || classifier !is KClass<*>) {
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt]: Can we test this code path?

File: src/main/java/com/squareup/kotlinpoet/TypeVariableName.kt [hunk 11]
```diff
@@ -148,3 +150,12 @@ fun TypeParameterElement.asTypeVariableName(): TypeVariableName {
   val boundsTypeNames = bounds.map { it.asTypeName() }
   return TypeVariableName.of(name, boundsTypeNames, variance = null)
 }
+
+fun KTypeParameter.asTypeVariableName(): TypeVariableName {
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/TypeVariableName.kt]: Are we able to test this function as well?

**Comment by JakeWharton** [comment]
This looks good. Just some small nits left I think.

File: src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt (hunk 10 above)
Comment (evgeny-pasynkov) [reply src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt]: According to Kt documentation: "Returns `null` if this type is not denotable in Kotlin, for example if it is an intersection type.". So building the artificial example to test this path is not worth it: anyway, the result is not presentable in Kotlin :)

**Comment by evgeny-pasynkov** [comment]
please could you merge this PR and publish new build to maven? I'd like to use this functionality

**Pull request merged**